design/one-pager-cross-resource-referencing.md

```diff
@@ -0,0 +1,296 @@
+  a `VPC`, since `VPC.Id` is required for provisioning the `Subnet`.
+
+  In addition we define `r2` has a *non-blocking* dependency on `r1`, if `r2`
+  doesn't require any attributes of `r1` to be provisioned but its functionality
+  assumes the existence of `r1`. For example, an `eks` cluster needs to have the
+  right `IAMRolePolicyAttachment` resource in order to have to required
+  permissions when accessing resources, even though it will be provisioned even
+  if the policy resource doesn't exist.
+
+* **Non-deterministic Resource attribute**
+
+  We define the attribute `foo` of resource `r` *non-deterministic*, if `foo`'s
+  value only becomes known after `r` is provisioned. For example `vpcID` is a
+  non-deterministic attribute of a `VPC` resource.
+
+* **Composite Resource Attribute**
+
+  We define the attribute `foo` of resource `r` *composite*, if `foo`'s value is
+  deterministic and composed of the attributes of other resources. As an
+  example, `network` attribute of `Subnetwork` type in GCP, is formed as
+
+  `/projects/[gcp-project]/global/networks/[network-name]`
+
+  where `[netowork-name]` is an attribute of `Network` resource.
+
+## Objectives
+
+### Goals
+
+* Support `gitops` style declarative resource provisioning. Apply a directory of
+  YAML resources, which will eventually become a series of online and
+  functioning managed resources. This requires:
+  * Support to reference attributes of other resources in a resource YAML object
+  * Support for cross resource blocking dependency
+
+* Support using existing external resources to be referenced by the YAML objects
+
+### Non Goals
+
+* Support non-blocking dependencies between resources
+
+## Proposal
+
+Let's consider the following sample solution where a `VPC`, and a `Subnet` needs to be
+provisioned. The YAML object will look like following:
+
+```yaml
+---
+apiVersion: network.aws.crossplane.io/v1alpha2
+kind: VPC
+metadata:
+  namespace: cool-ns
+  name: my-vpc
+spec:
+  ...
+---
+apiVersion: network.aws.crossplane.io/v1alpha2
+kind: Subnet
+metadata:
+  namespace: cool-ns
+  name: my-subnet
+spec:
+  # this is the vpcId of the external vpc, represented by my-vpc
+  vpcId: [my-vpc_vpcId]
+  ...
+```
+
+### Cross referencing using Reference Resolvers
+
+In this example since the `vpcId` is non-deterministic, `[my-vpc_vpcId]` is not
+known prior to provisioning the `VPC`. Therefore we will need a mechanism to
+indicate this cross reference in the YAML  object. To solve this, I propose the
+notion of *Reference Resolver*, as a `go` interface as following:
+
+```go
+type ReferenceResolver interface {
+  // resolves the attribute, or returns an error
+  Resolve() (string,error)
+}
+```
+
+Having this interface, we can implement `VpcIDRefResolver` as
+
+```go
+type VpcIDRefResolver struct {
+  // the list of objects that are needed for resolving vpcID
+  // in this case we only need VPC name, but we could reference more objects
+  VPCName string `json:"vpcName"`
```

Comment:
Are we making the assumption that references to a dependent resource would always be in the same namespace?

Reply:
I see below that it looks like we are. Are we certain that we want to impose this restriction?

Reply:
@hasheddan This is a simplifying restriction that we made at the first iteration. The main reason is that when cleaning things up, it would be easier to search for dependent resources if they were all in the same namespace (I just realized that I haven't added the clean-up section to this doc yet, coming up soon)

Reply:
@negz I see two issues with using `LocalObjectReference`:
- As you mentioned, we wouldn't be able to refer to more than one resource
- the `name`, for the field is a bit ambiguous. Note that if we use `LocalObjectReference`, we will have something like:

```yaml
spec:
  # reference to API objects from which the vpcId will be resolved
  vpcIdRef:
    name: my-vpc
```

where as there is no `vpcId` resource as implied by `name`.

Reply:
> Are we certain that we want to impose this restriction?

I'd rather look at this from the perspective of whether we expect to have use cases in which a managed resource in one namespace depends on a managed resource in another namespace. Given how we've predicated folks might use namespaces to logically group production infrastructure, and given what resources typically have hard dependencies on other resources, I struggle to think of many.

Reply:
> but would fall apart if we needed to reference more than one resource.

@negz I cannot think of a spec field value that depends on more than one resource, in the end 1 value will be fetched for 1 field and we can just get that from the referenced object. If we need more fields from the same resource, it could be just another reference like;
```
vpcIdRef *LocalObjectReference
vpcSomeOtherFieldNameRef *LocalObjectReference
```

Reply:
How about this struct
```go
type VpcIDRefResolver struct {
  // the list of objects that are needed for resolving vpcID
  // in this case we only need VPC name, but we could reference more objects
  VPC *LocalObjectReference `json:"vpc"`
}
```
and using it like 
```go
vpcIdRef *VPCIDRefResolver
```
Which the YAML will look like:
```yaml
vpcIdRef:
  vpc:
    name: my-vpc
```

Reply:
I don't think this really makes us compliant with what is suggested by the API conventions. Both this and the "vpcName" proposal mean we're using "something close to what Kubernetes recommends" rather than "what Kubernetes recommends". Possibly we get enough benefit from diverging from API conventions to make doing so worthwhile, but I don't personally think we do.